api/src/main/java/org/geysermc/geyser/api/item/custom/CustomItemData.java

```diff
@@ -114,6 +115,83 @@ public interface CustomItemData {
      */
     @NonNull Set<String> tags();
 
+    /**
+     * Gets the stack size of the item.
+     *
+     * Returns 0 if not set. When not set (or 0), it defaults to the stack count of the Java item when based of a vanilla item, or 64 when registering a non-vanilla item.
+     *
+     * @return the stack size of the item
+     */
+    @NonNegative
+    int stackSize();
+
+    /**
+     * Gets the max damage of the item.
+     *
+     * Returns -1 if not set. When not set (or below 0), it defaults to the maximum damage of the Java item when based of a vanilla item, or uses 0 when registering a non-vanilla item.
```

Comment:
Why are we defaulting to -1 instead of 0?

Reply:
Someone might want to create a custom vanilla item definition with a vanilla item that does have damage, but the item override doesn't. This allows for properly detecting that. E.g.:

```json
"minecraft:diamond_sword": [
  {
    "custom_model_data": 1,
    "display_name": "Diamond Sword override without damage!",
    "max_damage": 0,
    "name": "special_sword",
    "stack_size": 1
  }
]
```

Simple command to create such a sword on Java: `/give @s minecraft:diamond_sword[minecraft:custom_model_data=1,!minecraft:max_damage]`

If we defaulted to 0, then we can't be sure whether the value of 0 was implicitly or explicitly set and if we should use the Java base item damage value or not ([relevant code](https://github.com/GeyserMC/Geyser/blob/7881c4641195b1fc6940f6ff38d716c65d4f5034/core/src/main/java/org/geysermc/geyser/registry/populator/CustomItemRegistryPopulator.java#L169)).

Speaking of this - Java doesn't allow items to have a max damage value above 0 if the stack size is above 1 (and vice versa), in other words, items can't have durability if they are stackable. Should we implement this in the API, and have stack size default to 1 when setting max damage above 0 and max damage default to 0 when setting a stack size above 1?

Reply:
hmm, i see. Thanks for elaborating!
I would add the Java limit tbh, we do follow java standards in other places too

Reply:
The Java limit should be implemented now